File: 2d/ui/src/editor/editor_settings.js
export const EDITOR_SETTINGS_DEFAULTS = Object.freeze({
  unit: "cm",
  snapOn: true,
  snapCornerEnabled: true,
  snapMidEnabled: true,
  snapCenterEnabled: true,
  snapEdgeEnabled: true,
  wallMagnetEnabled: true,
  orthoEnabled: true,
  stepDrawMode: "line",
  stepDrawEnabled: true,
  stepLineEnabled: true,
  stepDegreeEnabled: false,
  stepLineCm: 5,
  stepAngleDeg: 10,
  dimSnapProfileEnabled: false,
  dimSnapLineProfile: { origin: true, corner: true, mid: true, center: true, edge: true },
  dimSnapOffsetProfile: { origin: true, corner: true, mid: true, center: true, edge: true },
  meterDivisions: 10,
  majorEvery: 10,
  bgColor: "#FFFFFF",
  minorColor: "#E6E6E6",
  majorColor: "#A3A3A3",
  axisXColor: "#9CC9B4",
  axisYColor: "#BCC8EB",
  axisZColor: "#0000FF",
  showObjectAxes: false,
  wallFillColor: "#A6A6A6",
  wallEdgeColor: "#000000",
  wallTextColor: "#FFFFFF",
  wallHeightColor: "#4B5563",
  wall3dColor: "#C7CCD1",
  beamThicknessMm: 400,
  beamHeightMm: 200,
  beamFloorOffsetMm: 2600,
  beamFillColor: "#A6A6A6",
  beamEdgeColor: "#000000",
  beamTextColor: "#FFFFFF",
  beam3dColor: "#C7CCD1",
  columnWidthMm: 500,
  columnDepthMm: 400,
  columnHeightMm: 2800,
  columnFillColor: "#A6A6A6",
  columnEdgeColor: "#000000",
  columnTextColor: "#FFFFFF",
  column3dColor: "#C7CCD1",
  wallThicknessMm: 120,
  wallHeightMm: 3000,
  hiddenWallThicknessMm: 1,
  hiddenWallColor: "#D8D4D4",
  hiddenWallLineWidthPx: 2,
  hiddenWallDash: [10, 8],
  dimColor: "#E8A559",
  dimFontPx: 15,
  dimLineWidthPx: 2,
  dimOffsetMm: 150,
  dimTickPx: 8,
  dimGapTextPx: 6,
  dimTextPadPx: 6,
  showDimensions: true,
  angleColor: "#E8A559",
  angleFontPx: 12,
  angleRadiusPx: 22,
  angleArcWidthPx: 2,
  offsetWallEnabled: true,
  offsetWallDistanceMm: 600,
  offsetWallColor: "#D8D4D4",
  offsetWallLineWidthPx: 2,
  offsetWallDash: [10, 8],
  fontFamily: "Tahoma",
  wallNameFontPx: 15,
  miterLimit: 10,
  showGammaDebug: false,
});

function cloneProfile(value, fallback) {
  const source = value && typeof value === "object" ? value : fallback;
  return {
    origin: source.origin !== false,
    corner: source.corner !== false,
    mid: source.mid !== false,
    center: source.center !== false,
    edge: source.edge !== false,
  };
}

function cloneDash(value, fallback) {
  const source = Array.isArray(value) && value.length ? value : fallback;
  return source
    .map((item) => Math.max(0, Number(item) || 0))
    .filter((item) => Number.isFinite(item));
}

function normalizeString(value, fallback) {
  const text = String(value ?? "").trim();
  return text || fallback;
}

function normalizeBool(value, fallback) {
  return typeof value === "boolean" ? value : fallback;
}

function normalizeNumber(value, fallback, min = null) {
  const parsed = Number(value);
  if (!Number.isFinite(parsed)) return fallback;
  if (min != null && parsed < min) return fallback;
  return parsed;
}

export function normalizeEditorSettingsState(input = {}) {
  const source = input && typeof input === "object" ? input : {};
  return {
    unit: normalizeString(source.unit, EDITOR_SETTINGS_DEFAULTS.unit),
    snapOn: normalizeBool(source.snapOn, EDITOR_SETTINGS_DEFAULTS.snapOn),
    snapCornerEnabled: normalizeBool(source.snapCornerEnabled, EDITOR_SETTINGS_DEFAULTS.snapCornerEnabled),
    snapMidEnabled: normalizeBool(source.snapMidEnabled, EDITOR_SETTINGS_DEFAULTS.snapMidEnabled),
    snapCenterEnabled: normalizeBool(source.snapCenterEnabled, EDITOR_SETTINGS_DEFAULTS.snapCenterEnabled),
    snapEdgeEnabled: normalizeBool(source.snapEdgeEnabled, EDITOR_SETTINGS_DEFAULTS.snapEdgeEnabled),
    wallMagnetEnabled: normalizeBool(source.wallMagnetEnabled, EDITOR_SETTINGS_DEFAULTS.wallMagnetEnabled),
    orthoEnabled: normalizeBool(source.orthoEnabled, EDITOR_SETTINGS_DEFAULTS.orthoEnabled),
    stepDrawMode: normalizeString(source.stepDrawMode, EDITOR_SETTINGS_DEFAULTS.stepDrawMode) === "degree" ? "degree" : "line",
    stepDrawEnabled: normalizeBool(source.stepDrawEnabled, EDITOR_SETTINGS_DEFAULTS.stepDrawEnabled),
    stepLineEnabled: normalizeBool(source.stepLineEnabled, EDITOR_SETTINGS_DEFAULTS.stepLineEnabled),
    stepDegreeEnabled: normalizeBool(source.stepDegreeEnabled, EDITOR_SETTINGS_DEFAULTS.stepDegreeEnabled),
    stepLineCm: normalizeNumber(source.stepLineCm, EDITOR_SETTINGS_DEFAULTS.stepLineCm, 0.1),
    stepAngleDeg: normalizeNumber(source.stepAngleDeg, EDITOR_SETTINGS_DEFAULTS.stepAngleDeg, 0.1),
    dimSnapProfileEnabled: normalizeBool(source.dimSnapProfileEnabled, EDITOR_SETTINGS_DEFAULTS.dimSnapProfileEnabled),
    dimSnapLineProfile: cloneProfile(source.dimSnapLineProfile, EDITOR_SETTINGS_DEFAULTS.dimSnapLineProfile),
    dimSnapOffsetProfile: cloneProfile(source.dimSnapOffsetProfile, EDITOR_SETTINGS_DEFAULTS.dimSnapOffsetProfile),
    meterDivisions: normalizeNumber(source.meterDivisions, EDITOR_SETTINGS_DEFAULTS.meterDivisions, 1),
    majorEvery: normalizeNumber(source.majorEvery, EDITOR_SETTINGS_DEFAULTS.majorEvery, 1),
    bgColor: normalizeString(source.bgColor, EDITOR_SETTINGS_DEFAULTS.bgColor),
    minorColor: normalizeString(source.minorColor, EDITOR_SETTINGS_DEFAULTS.minorColor),
    majorColor: normalizeString(source.majorColor, EDITOR_SETTINGS_DEFAULTS.majorColor),
    axisXColor: normalizeString(source.axisXColor, EDITOR_SETTINGS_DEFAULTS.axisXColor),
    axisYColor: normalizeString(source.axisYColor, EDITOR_SETTINGS_DEFAULTS.axisYColor),
    axisZColor: normalizeString(source.axisZColor, EDITOR_SETTINGS_DEFAULTS.axisZColor),
    showObjectAxes: normalizeBool(source.showObjectAxes, EDITOR_SETTINGS_DEFAULTS.showObjectAxes),
    wallFillColor: normalizeString(source.wallFillColor, EDITOR_SETTINGS_DEFAULTS.wallFillColor),
    wallEdgeColor: normalizeString(source.wallEdgeColor, EDITOR_SETTINGS_DEFAULTS.wallEdgeColor),
    wallTextColor: normalizeString(source.wallTextColor, EDITOR_SETTINGS_DEFAULTS.wallTextColor),
    wallHeightColor: normalizeString(source.wallHeightColor, EDITOR_SETTINGS_DEFAULTS.wallHeightColor),
    wall3dColor: normalizeString(source.wall3dColor, EDITOR_SETTINGS_DEFAULTS.wall3dColor),
    beamThicknessMm: normalizeNumber(source.beamThicknessMm, EDITOR_SETTINGS_DEFAULTS.beamThicknessMm, 1),
    beamHeightMm: normalizeNumber(source.beamHeightMm, EDITOR_SETTINGS_DEFAULTS.beamHeightMm, 1),
    beamFloorOffsetMm: normalizeNumber(source.beamFloorOffsetMm, EDITOR_SETTINGS_DEFAULTS.beamFloorOffsetMm, 0),
    beamFillColor: normalizeString(source.beamFillColor, EDITOR_SETTINGS_DEFAULTS.beamFillColor),
    beamEdgeColor: normalizeString(source.beamEdgeColor, EDITOR_SETTINGS_DEFAULTS.beamEdgeColor),
    beamTextColor: normalizeString(source.beamTextColor, EDITOR_SETTINGS_DEFAULTS.beamTextColor),
    beam3dColor: normalizeString(source.beam3dColor, EDITOR_SETTINGS_DEFAULTS.beam3dColor),
    columnWidthMm: normalizeNumber(source.columnWidthMm, EDITOR_SETTINGS_DEFAULTS.columnWidthMm, 1),
    columnDepthMm: normalizeNumber(source.columnDepthMm, EDITOR_SETTINGS_DEFAULTS.columnDepthMm, 1),
    columnHeightMm: normalizeNumber(source.columnHeightMm, EDITOR_SETTINGS_DEFAULTS.columnHeightMm, 1),
    columnFillColor: normalizeString(source.columnFillColor, EDITOR_SETTINGS_DEFAULTS.columnFillColor),
    columnEdgeColor: normalizeString(source.columnEdgeColor, EDITOR_SETTINGS_DEFAULTS.columnEdgeColor),
    columnTextColor: normalizeString(source.columnTextColor, EDITOR_SETTINGS_DEFAULTS.columnTextColor),
    column3dColor: normalizeString(source.column3dColor, EDITOR_SETTINGS_DEFAULTS.column3dColor),
    wallThicknessMm: normalizeNumber(source.wallThicknessMm, EDITOR_SETTINGS_DEFAULTS.wallThicknessMm, 1),
    wallHeightMm: normalizeNumber(source.wallHeightMm, EDITOR_SETTINGS_DEFAULTS.wallHeightMm, 1),
    hiddenWallThicknessMm: normalizeNumber(source.hiddenWallThicknessMm, EDITOR_SETTINGS_DEFAULTS.hiddenWallThicknessMm, 0.1),
    hiddenWallColor: normalizeString(source.hiddenWallColor, EDITOR_SETTINGS_DEFAULTS.hiddenWallColor),
    hiddenWallLineWidthPx: normalizeNumber(source.hiddenWallLineWidthPx, EDITOR_SETTINGS_DEFAULTS.hiddenWallLineWidthPx, 1),
    hiddenWallDash: cloneDash(source.hiddenWallDash, EDITOR_SETTINGS_DEFAULTS.hiddenWallDash),
    dimColor: normalizeString(source.dimColor, EDITOR_SETTINGS_DEFAULTS.dimColor),
    dimFontPx: normalizeNumber(source.dimFontPx, EDITOR_SETTINGS_DEFAULTS.dimFontPx, 1),
    dimLineWidthPx: normalizeNumber(source.dimLineWidthPx, EDITOR_SETTINGS_DEFAULTS.dimLineWidthPx, 1),
    dimOffsetMm: normalizeNumber(source.dimOffsetMm, EDITOR_SETTINGS_DEFAULTS.dimOffsetMm, 0),
    dimTickPx: normalizeNumber(source.dimTickPx, EDITOR_SETTINGS_DEFAULTS.dimTickPx, 1),
    dimGapTextPx: normalizeNumber(source.dimGapTextPx, EDITOR_SETTINGS_DEFAULTS.dimGapTextPx, 0),
    dimTextPadPx: normalizeNumber(source.dimTextPadPx, EDITOR_SETTINGS_DEFAULTS.dimTextPadPx, 0),
    showDimensions: normalizeBool(source.showDimensions, EDITOR_SETTINGS_DEFAULTS.showDimensions),
    angleColor: normalizeString(source.angleColor, EDITOR_SETTINGS_DEFAULTS.angleColor),
    angleFontPx: normalizeNumber(source.angleFontPx, EDITOR_SETTINGS_DEFAULTS.angleFontPx, 1),
    angleRadiusPx: normalizeNumber(source.angleRadiusPx, EDITOR_SETTINGS_DEFAULTS.angleRadiusPx, 1),
    angleArcWidthPx: normalizeNumber(source.angleArcWidthPx, EDITOR_SETTINGS_DEFAULTS.angleArcWidthPx, 1),
    offsetWallEnabled: normalizeBool(source.offsetWallEnabled, EDITOR_SETTINGS_DEFAULTS.offsetWallEnabled),
    offsetWallDistanceMm: normalizeNumber(source.offsetWallDistanceMm, EDITOR_SETTINGS_DEFAULTS.offsetWallDistanceMm, 0),
    offsetWallColor: normalizeString(source.offsetWallColor, EDITOR_SETTINGS_DEFAULTS.offsetWallColor),
    offsetWallLineWidthPx: normalizeNumber(source.offsetWallLineWidthPx, EDITOR_SETTINGS_DEFAULTS.offsetWallLineWidthPx, 1),
    offsetWallDash: cloneDash(source.offsetWallDash, EDITOR_SETTINGS_DEFAULTS.offsetWallDash),
    fontFamily: normalizeString(source.fontFamily, EDITOR_SETTINGS_DEFAULTS.fontFamily),
    wallNameFontPx: normalizeNumber(source.wallNameFontPx, EDITOR_SETTINGS_DEFAULTS.wallNameFontPx, 1),
    miterLimit: normalizeNumber(source.miterLimit, EDITOR_SETTINGS_DEFAULTS.miterLimit, 1),
    showGammaDebug: normalizeBool(source.showGammaDebug, EDITOR_SETTINGS_DEFAULTS.showGammaDebug),
  };
}

export function editorSettingsStateToPayload(input = {}) {
  const state = normalizeEditorSettingsState(input);
  return {
    generalSettings: {
      unit: state.unit,
      fontFamily: state.fontFamily,
      wallNameFontPx: state.wallNameFontPx,
    },
    gridSettings: {
      meterDivisions: state.meterDivisions,
      majorEvery: state.majorEvery,
      bgColor: state.bgColor,
      minorColor: state.minorColor,
      majorColor: state.majorColor,
      axisXColor: state.axisXColor,
      axisYColor: state.axisYColor,
      axisZColor: state.axisZColor,
    },
    snapSettings: {
      snapOn: state.snapOn,
      snapCornerEnabled: state.snapCornerEnabled,
      snapMidEnabled: state.snapMidEnabled,
      snapCenterEnabled: state.snapCenterEnabled,
      snapEdgeEnabled: state.snapEdgeEnabled,
      wallMagnetEnabled: state.wallMagnetEnabled,
      orthoEnabled: state.orthoEnabled,
      dimSnapProfileEnabled: state.dimSnapProfileEnabled,
      dimSnapLineProfile: cloneProfile(state.dimSnapLineProfile, EDITOR_SETTINGS_DEFAULTS.dimSnapLineProfile),
      dimSnapOffsetProfile: cloneProfile(state.dimSnapOffsetProfile, EDITOR_SETTINGS_DEFAULTS.dimSnapOffsetProfile),
    },
    draftingSettings: {
      stepDrawMode: state.stepDrawMode,
      stepDrawEnabled: state.stepDrawEnabled,
      stepLineEnabled: state.stepLineEnabled,
      stepDegreeEnabled: state.stepDegreeEnabled,
      stepLineCm: state.stepLineCm,
      stepAngleDeg: state.stepAngleDeg,
      showObjectAxes: state.showObjectAxes,
      miterLimit: state.miterLimit,
      showGammaDebug: state.showGammaDebug,
    },
    wallDefaults: {
      wallThicknessMm: state.wallThicknessMm,
      wallHeightMm: state.wallHeightMm,
      wallFillColor: state.wallFillColor,
      wallEdgeColor: state.wallEdgeColor,
      wallTextColor: state.wallTextColor,
      wallHeightColor: state.wallHeightColor,
      wall3dColor: state.wall3dColor,
    },
    beamDefaults: {
      beamThicknessMm: state.beamThicknessMm,
      beamHeightMm: state.beamHeightMm,
      beamFloorOffsetMm: state.beamFloorOffsetMm,
      beamFillColor: state.beamFillColor,
      beamEdgeColor: state.beamEdgeColor,
      beamTextColor: state.beamTextColor,
      beam3dColor: state.beam3dColor,
    },
    columnDefaults: {
      columnWidthMm: state.columnWidthMm,
      columnDepthMm: state.columnDepthMm,
      columnHeightMm: state.columnHeightMm,
      columnFillColor: state.columnFillColor,
      columnEdgeColor: state.columnEdgeColor,
      columnTextColor: state.columnTextColor,
      column3dColor: state.column3dColor,
    },
    hiddenDefaults: {
      hiddenWallThicknessMm: state.hiddenWallThicknessMm,
      hiddenWallColor: state.hiddenWallColor,
      hiddenWallLineWidthPx: state.hiddenWallLineWidthPx,
      hiddenWallDash: cloneDash(state.hiddenWallDash, EDITOR_SETTINGS_DEFAULTS.hiddenWallDash),
    },
    dimensionDefaults: {
      dimColor: state.dimColor,
      dimFontPx: state.dimFontPx,
      dimLineWidthPx: state.dimLineWidthPx,
      dimOffsetMm: state.dimOffsetMm,
      dimTickPx: state.dimTickPx,
      dimGapTextPx: state.dimGapTextPx,
      dimTextPadPx: state.dimTextPadPx,
      showDimensions: state.showDimensions,
    },
    angleDefaults: {
      angleColor: state.angleColor,
      angleFontPx: state.angleFontPx,
      angleRadiusPx: state.angleRadiusPx,
      angleArcWidthPx: state.angleArcWidthPx,
    },
    offsetWallSettings: {
      offsetWallEnabled: state.offsetWallEnabled,
      offsetWallDistanceMm: state.offsetWallDistanceMm,
      offsetWallColor: state.offsetWallColor,
      offsetWallLineWidthPx: state.offsetWallLineWidthPx,
      offsetWallDash: cloneDash(state.offsetWallDash, EDITOR_SETTINGS_DEFAULTS.offsetWallDash),
    },
  };
}

export function editorSettingsPayloadToState(payload = {}) {
  const source = payload && typeof payload === "object" ? payload : {};
  return normalizeEditorSettingsState({
    ...EDITOR_SETTINGS_DEFAULTS,
    ...(source.generalSettings || {}),
    ...(source.gridSettings || {}),
    ...(source.snapSettings || {}),
    ...(source.draftingSettings || {}),
    ...(source.wallDefaults || {}),
    ...(source.beamDefaults || {}),
    ...(source.columnDefaults || {}),
    ...(source.hiddenDefaults || {}),
    ...(source.dimensionDefaults || {}),
    ...(source.angleDefaults || {}),
    ...(source.offsetWallSettings || {}),
  });
}

export function settingsViewStateFromEditorState(state = {}) {
  const flat = normalizeEditorSettingsState(state);
  return {
    unit: flat.unit,
    wallThicknessMm: flat.wallThicknessMm,
    wallHeightMm: flat.wallHeightMm,
    hiddenWallThicknessMm: flat.hiddenWallThicknessMm,
    beamThicknessMm: flat.beamThicknessMm,
    beamHeightMm: flat.beamHeightMm,
    beamFloorOffsetMm: flat.beamFloorOffsetMm,
    columnWidthMm: flat.columnWidthMm,
    columnDepthMm: flat.columnDepthMm,
    columnHeightMm: flat.columnHeightMm,
    dimOffsetMm: flat.dimOffsetMm,
    dimFontPx: flat.dimFontPx,
    dimLineWidthPx: flat.dimLineWidthPx,
    meterDivisions: flat.meterDivisions,
    majorEvery: flat.majorEvery,
    fontFamily: flat.fontFamily,
    wallNameFontPx: flat.wallNameFontPx,
    angleFontPx: flat.angleFontPx,
    snapOn: flat.snapOn,
    showDimensions: flat.showDimensions,
    stepDrawEnabled: flat.stepDrawEnabled,
    stepLineCm: flat.stepLineCm,
    stepAngleDeg: flat.stepAngleDeg,
    snapCornerEnabled: flat.snapCornerEnabled,
    snapMidEnabled: flat.snapMidEnabled,
    snapCenterEnabled: flat.snapCenterEnabled,
    snapEdgeEnabled: flat.snapEdgeEnabled,
    wallMagnetEnabled: flat.wallMagnetEnabled,
    bgColor: flat.bgColor,
    minorColor: flat.minorColor,
    majorColor: flat.majorColor,
    axisXColor: flat.axisXColor,
    axisYColor: flat.axisYColor,
    axisZColor: flat.axisZColor,
    wallFillColor: flat.wallFillColor,
    wallEdgeColor: flat.wallEdgeColor,
    wallTextColor: flat.wallTextColor,
    wall3dColor: flat.wall3dColor,
    beamFillColor: flat.beamFillColor,
    beamEdgeColor: flat.beamEdgeColor,
    beamTextColor: flat.beamTextColor,
    beam3dColor: flat.beam3dColor,
    columnFillColor: flat.columnFillColor,
    columnEdgeColor: flat.columnEdgeColor,
    columnTextColor: flat.columnTextColor,
    column3dColor: flat.column3dColor,
    dimColor: flat.dimColor,
    hiddenWallColor: flat.hiddenWallColor,
    offsetWallEnabled: flat.offsetWallEnabled,
    offsetWallDistanceMm: flat.offsetWallDistanceMm,
  };
}

export async function fetchPersistedEditorSettings(adminId, userId) {
  const res = await fetch(
    `/api/editor-settings?admin_id=${encodeURIComponent(adminId)}&user_id=${encodeURIComponent(userId)}`
  );
  if (!res.ok) {
    throw new Error("editor-settings-fetch-failed");
  }
  return await res.json();
}

export async function savePersistedEditorSettings(adminId, userId, payload) {
  const res = await fetch(
    `/api/editor-settings?admin_id=${encodeURIComponent(adminId)}&user_id=${encodeURIComponent(userId)}`,
    {
      method: "PUT",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(payload),
    }
  );
  if (!res.ok) {
    throw new Error("editor-settings-save-failed");
  }
  return await res.json();
}

File: backend/src/designkp_backend/api/routers/editor_settings.py
from __future__ import annotations

import uuid
from datetime import datetime

from fastapi import APIRouter, Depends, HTTPException, Query, status
from pydantic import BaseModel, Field
from sqlalchemy import and_, select
from sqlalchemy.ext.asyncio import AsyncSession

from designkp_backend.db.dependencies import get_db_session
from designkp_backend.db.models.account import EditorSetting, User
from designkp_backend.services.admin_access import require_admin

router = APIRouter(prefix="/editor-settings", tags=["editor_settings"])


class SnapProfile(BaseModel):
    origin: bool = True
    corner: bool = True
    mid: bool = True
    center: bool = True
    edge: bool = True


class GeneralSettingsPayload(BaseModel):
    unit: str = "cm"
    fontFamily: str = "Tahoma"
    wallNameFontPx: int = 15


class GridSettingsPayload(BaseModel):
    meterDivisions: int = 10
    majorEvery: int = 10
    bgColor: str = "#FFFFFF"
    minorColor: str = "#E6E6E6"
    majorColor: str = "#A3A3A3"
    axisXColor: str = "#9CC9B4"
    axisYColor: str = "#BCC8EB"
    axisZColor: str = "#0000FF"


class SnapSettingsPayload(BaseModel):
    snapOn: bool = True
    snapCornerEnabled: bool = True
    snapMidEnabled: bool = True
    snapCenterEnabled: bool = True
    snapEdgeEnabled: bool = True
    wallMagnetEnabled: bool = True
    orthoEnabled: bool = True
    dimSnapProfileEnabled: bool = False
    dimSnapLineProfile: SnapProfile = Field(default_factory=SnapProfile)
    dimSnapOffsetProfile: SnapProfile = Field(default_factory=SnapProfile)


class DraftingSettingsPayload(BaseModel):
    stepDrawMode: str = "line"
    stepDrawEnabled: bool = True
    stepLineEnabled: bool = True
    stepDegreeEnabled: bool = False
    stepLineCm: float = 5
    stepAngleDeg: float = 10
    showObjectAxes: bool = False
    miterLimit: int = 10
    showGammaDebug: bool = False


class WallDefaultsPayload(BaseModel):
    wallThicknessMm: int = 120
    wallHeightMm: int = 3000
    wallFillColor: str = "#A6A6A6"
    wallEdgeColor: str = "#000000"
    wallTextColor: str = "#FFFFFF"
    wallHeightColor: str = "#4B5563"
    wall3dColor: str = "#C7CCD1"


class BeamDefaultsPayload(BaseModel):
    beamThicknessMm: int = 400
    beamHeightMm: int = 200
    beamFloorOffsetMm: int = 2600
    beamFillColor: str = "#A6A6A6"
    beamEdgeColor: str = "#000000"
    beamTextColor: str = "#FFFFFF"
    beam3dColor: str = "#C7CCD1"


class ColumnDefaultsPayload(BaseModel):
    columnWidthMm: int = 500
    columnDepthMm: int = 400
    columnHeightMm: int = 2800
    columnFillColor: str = "#A6A6A6"
    columnEdgeColor: str = "#000000"
    columnTextColor: str = "#FFFFFF"
    column3dColor: str = "#C7CCD1"


class HiddenDefaultsPayload(BaseModel):
    hiddenWallThicknessMm: float = 1
    hiddenWallColor: str = "#D8D4D4"
    hiddenWallLineWidthPx: int = 2
    hiddenWallDash: list[int] = Field(default_factory=lambda: [10, 8])


class DimensionDefaultsPayload(BaseModel):
    dimColor: str = "#E8A559"
    dimFontPx: int = 15
    dimLineWidthPx: int = 2
    dimOffsetMm: int = 150
    dimTickPx: int = 8
    dimGapTextPx: int = 6
    dimTextPadPx: int = 6
    showDimensions: bool = True


class AngleDefaultsPayload(BaseModel):
    angleColor: str = "#E8A559"
    angleFontPx: int = 12
    angleRadiusPx: int = 22
    angleArcWidthPx: int = 2


class OffsetWallSettingsPayload(BaseModel):
    offsetWallEnabled: bool = True
    offsetWallDistanceMm: int = 600
    offsetWallColor: str = "#D8D4D4"
    offsetWallLineWidthPx: int = 2
    offsetWallDash: list[int] = Field(default_factory=lambda: [10, 8])


class EditorSettingsPayload(BaseModel):
    generalSettings: GeneralSettingsPayload = Field(default_factory=GeneralSettingsPayload)
    gridSettings: GridSettingsPayload = Field(default_factory=GridSettingsPayload)
    snapSettings: SnapSettingsPayload = Field(default_factory=SnapSettingsPayload)
    draftingSettings: DraftingSettingsPayload = Field(default_factory=DraftingSettingsPayload)
    wallDefaults: WallDefaultsPayload = Field(default_factory=WallDefaultsPayload)
    beamDefaults: BeamDefaultsPayload = Field(default_factory=BeamDefaultsPayload)
    columnDefaults: ColumnDefaultsPayload = Field(default_factory=ColumnDefaultsPayload)
    hiddenDefaults: HiddenDefaultsPayload = Field(default_factory=HiddenDefaultsPayload)
    dimensionDefaults: DimensionDefaultsPayload = Field(default_factory=DimensionDefaultsPayload)
    angleDefaults: AngleDefaultsPayload = Field(default_factory=AngleDefaultsPayload)
    offsetWallSettings: OffsetWallSettingsPayload = Field(default_factory=OffsetWallSettingsPayload)


class EditorSettingsItem(EditorSettingsPayload):
    id: uuid.UUID | None = None
    admin_id: uuid.UUID
    user_id: uuid.UUID
    updated_at: datetime | None = None
    persisted: bool = False


async def _require_accessible_user(session: AsyncSession, *, admin_id: uuid.UUID, user_id: uuid.UUID) -> User:
    user = await session.scalar(
        select(User).where(
            and_(
                User.id == user_id,
                User.admin_id == admin_id,
                User.deleted_at.is_(None),
            )
        )
    )
    if not user:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="User not found for this admin.")
    return user


def _default_payload() -> EditorSettingsPayload:
    return EditorSettingsPayload()


def _to_response(item: EditorSetting | None, *, admin_id: uuid.UUID, user_id: uuid.UUID) -> EditorSettingsItem:
    if item is None:
        default_payload = _default_payload()
        return EditorSettingsItem(
            id=None,
            admin_id=admin_id,
            user_id=user_id,
            updated_at=None,
            persisted=False,
            **default_payload.model_dump(),
        )

    normalized = EditorSettingsPayload(
        generalSettings=GeneralSettingsPayload.model_validate(item.general_settings or {}),
        gridSettings=GridSettingsPayload.model_validate(item.grid_settings or {}),
        snapSettings=SnapSettingsPayload.model_validate(item.snap_settings or {}),
        draftingSettings=DraftingSettingsPayload.model_validate(item.drafting_settings or {}),
        wallDefaults=WallDefaultsPayload.model_validate(item.wall_defaults or {}),
        beamDefaults=BeamDefaultsPayload.model_validate(item.beam_defaults or {}),
        columnDefaults=ColumnDefaultsPayload.model_validate(item.column_defaults or {}),
        hiddenDefaults=HiddenDefaultsPayload.model_validate(item.hidden_defaults or {}),
        dimensionDefaults=DimensionDefaultsPayload.model_validate(item.dimension_defaults or {}),
        angleDefaults=AngleDefaultsPayload.model_validate(item.angle_defaults or {}),
        offsetWallSettings=OffsetWallSettingsPayload.model_validate(item.offset_wall_settings or {}),
    )
    return EditorSettingsItem(
        id=item.id,
        admin_id=item.admin_id,
        user_id=item.user_id,
        updated_at=item.updated_at,
        persisted=True,
        **normalized.model_dump(),
    )


@router.get("", response_model=EditorSettingsItem)
async def get_editor_settings(
    admin_id: uuid.UUID = Query(...),
    user_id: uuid.UUID = Query(...),
    session: AsyncSession = Depends(get_db_session),
) -> EditorSettingsItem:
    await require_admin(session, admin_id)
    await _require_accessible_user(session, admin_id=admin_id, user_id=user_id)
    item = await session.scalar(
        select(EditorSetting).where(
            and_(
                EditorSetting.admin_id == admin_id,
                EditorSetting.user_id == user_id,
                EditorSetting.deleted_at.is_(None),
            )
        )
    )
    return _to_response(item, admin_id=admin_id, user_id=user_id)


@router.put("", response_model=EditorSettingsItem)
async def upsert_editor_settings(
    payload: EditorSettingsPayload,
    admin_id: uuid.UUID = Query(...),
    user_id: uuid.UUID = Query(...),
    session: AsyncSession = Depends(get_db_session),
) -> EditorSettingsItem:
    await require_admin(session, admin_id)
    await _require_accessible_user(session, admin_id=admin_id, user_id=user_id)
    item = await session.scalar(
        select(EditorSetting).where(
            and_(
                EditorSetting.admin_id == admin_id,
                EditorSetting.user_id == user_id,
            )
        )
    )
    if item is None:
        item = EditorSetting(admin_id=admin_id, user_id=user_id)
        session.add(item)

    item.deleted_at = None
    item.general_settings = payload.generalSettings.model_dump()
    item.grid_settings = payload.gridSettings.model_dump()
    item.snap_settings = payload.snapSettings.model_dump()
    item.drafting_settings = payload.draftingSettings.model_dump()
    item.wall_defaults = payload.wallDefaults.model_dump()
    item.beam_defaults = payload.beamDefaults.model_dump()
    item.column_defaults = payload.columnDefaults.model_dump()
    item.hidden_defaults = payload.hiddenDefaults.model_dump()
    item.dimension_defaults = payload.dimensionDefaults.model_dump()
    item.angle_defaults = payload.angleDefaults.model_dump()
    item.offset_wall_settings = payload.offsetWallSettings.model_dump()

    await session.commit()
    await session.refresh(item)
    return _to_response(item, admin_id=admin_id, user_id=user_id)
